src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,260 @@
+                .filter(name -> !new TurboLabel("", name).getGroupName().equals(group))
+                .collect(Collectors.toList());
+
+        // Add to removedLabels all initialLabels that have conflicting group
+        removedLabels.addAll(initialLabels.stream()
+                .filter(name -> new TurboLabel("", name).getGroupName().equals(group) 
+                    && !removedLabels.contains(label.getFullName()))
+                .collect(Collectors.toList()));
+    }
+
+    private boolean isAnInitialLabel(String name) {
+        return this.initialLabels.contains(name);
+    }
+
+    private boolean isAnAddedLabel(String name) {
+        return this.addedLabels.contains(name);
+    }
+
+    private boolean isARemovedLabel(String name) {
+        return this.removedLabels.contains(name);
+    }
+
+    /**
+     * @return suggested label 
+     */
+    private String getSuggestedLabel() {
+        return matchedLabels.get(currentSuggestionIndex.getAsInt());
+    }
+
+    /**
+     * @return true if a suggestion index is positive and within the list of matched labels
+     */
+    private boolean isValidSuggestionIndex() {
+        return currentSuggestionIndex.getAsInt() >= 0 && currentSuggestionIndex.getAsInt() < matchedLabels.size();
+    }
+
+
+    /**
+     * @param userInput
+     * @return list of confirmed keywords based on given userInput
+     */
+    private static List<String> getConfirmedKeywords(String userInput) {
+        ArrayList<String> confirmedKeywords = new ArrayList<>();
+
+        String[] keywords = userInput.split("\\s+");
+        for (int i = 0; i < keywords.length; i++) {
+            if (isConfirmedKeyword(userInput, i)) {
+                confirmedKeywords.add(keywords[i]);
+            }
+        }
+
+        return confirmedKeywords;
+    }
+
+    /**
+     * A keyword is considered in progress if there is no space after it i.e  given
+     * an input "test dummy world", "world" is the keyword in progress.
+     * @param userInput
+     * @return the keyword in progress based on the userInput
+     */
+    private static Optional<String> getKeywordInProgress(String userInput) {
+        String[] keywords = userInput.split("\\s+");
+        if (keywords.length == 0) return Optional.empty();
+
+        if (isConfirmedKeyword(userInput, keywords.length - 1)) return Optional.empty();
+
+        return Optional.of(keywords[keywords.length - 1]);
+    }
+
+    /**
+     * Determines if the keywordIndex-th keyword is confirmed i.e. user has typed a space after it
+     * Assumption: userInput has at at least keywordIndex+1 keywords, separated by whitespace.
+     * @param userInput
+     * @param keywordIndex
+     * @return
+     */
+    private static boolean isConfirmedKeyword(String userInput, int keywordIndex) {
+        return !(keywordIndex == userInput.split("\\s+").length - 1 && !userInput.endsWith(" "));
+    }
+
+    private static List<String> convertToList(Set<String> labelSet){
```

Comment:
Maybe this method can be removed now since it is rather simple and is only used once (in `getInitialLabels`)?